<center><img src="images/logo.png" alt="drawing" width="400" style="background-color:white; padding:1em;" /></center> <br/>

# ML through Application
## Module 1, Lab 3: Getting Started with AutoGluon

This notebook covers how to create a model to solve an ML problem by using [AutoGluon](https://auto.gluon.ai/stable/index.html#).

You will learn how to do the following:

- Import the AutoGluon library.
- Import data to a Pandas DataFrame.
- Train a model by using AutoGluon.

---

You will explore a dataset that contains information about books. The goal is to predict book prices by using features about the books.

__Business problem:__ Books from a large database with several features cannot be listed for sale because one critical piece of information is missing: the price. 

__ML problem description:__ Predict book prices by using book features, such as genre, release data, ratings, and number of reviews.

This is a regression task (the training dataset has a book price column to use for labels).

----

You will be presented with two kinds of exercises throughout the notebook: activities and challenges. <br/>

| <img style="float: center;" src="images/activity.png" alt="Activity" width="125"/>| <img style="float: center;" src="images/challenge.png" alt="Challenge" width="125"/>|
| --- | --- |
|<p style="text-align:center;">No coding is needed for an activity. You try to understand a concept, <br/>answer questions, or run a code cell.</p> |<p style="text-align:center;">Challenges are where you can practice your coding skills.</p>

## Index
- [Importing AutoGluon](#Importing-AutoGluon)
- [Getting the data](#Getting-the-data)
- [Model training with AutoGluon](#Model-training-with-AutoGluon)

---
## Importing AutoGluon

Install and load the libraries that are needed to work with the tabular dataset.

In [1]:
%%capture
# Use pip to install libraries
!pip install -U -q -r requirements.txt

In [2]:
# Import the libraries that are needed for the notebook
%load_ext autoreload
import pandas as pd
# Import utility functions and challenge questions
from MLUMLA_EN_M1_Lab3_quiz_questions import *

# Import the newly installed AutoGluon code library
from autogluon.tabular import TabularPredictor, TabularDataset

---
## Getting the data

Now get the data for the business problem. 

__Note:__ You will use the [Amazon Product Reviews](https://cseweb.ucsd.edu/~jmcauley/datasets.html#amazon_reviews) dataset. For more information about this dataset, see the following resources:

- Ruining He and Julian McAuley. "Ups and Downs: Modeling the Visual Evolution of Fashion Trends with One-Class Collaborative Filtering." Proceedings of the 25th International Conference on World Wide Web, Geneva, Switzerland, April 2016. https://doi.org/10.1145/2872427.2883037.

- Julian McAuley, Christopher Targett, Qinfeng Shi, Anton van den Hengel. "Image-Based Recommendations on Styles and Substitutes." Proceedings of the 38th International Association for Computing Machinery (ACM) Special Interest Group on Information Retrieval (SIGIR) Conference on Research and Development in Information Retrieval, Santiago, Chile, August 2015. https://doi.org/10.1145/2766462.2767755.

To load the training and test data, and then show the first few rows of the training dataset, run the following cells.

In [3]:
df_train = TabularDataset(data="data/train.csv")
df_test = TabularDataset(data="data/test.csv")

In [4]:
df_train.head()

category  \
0                                                 []   
1  ['Books', 'Education & Teaching', 'Schools & T...   
2                                                 []   
3                                                 []   
4                       ['Books', 'History', 'Asia']   

                                               title  \
0                                          Books" />   
1  The Core Knowledge Sequence Content and Skill ...   
2                              Stranger In The Woods   
3    Hansel and Gretel :  A Fairy Opera, Vocal Score   
4              Genghis Khan - Conqueror Of The World   

                                            also_buy  \
0                                                 []   
1  ['0325008957', '1138188492', '1890517208', '14...   
2                                                 []   
3                                                 []   
4                                                 []   

                                               brand                   rank  \
0                                      Joan M. Lexau   1,683,587 in Books (   
1                          Core Knowledge Foundation     974,014 in Books (   
2                                         Leah Fried  17,588,750 in Books (   
3  Adelheid ; Bache, Constance ; Humperdinck, E. ...   3,680,123 in Books (   
4                                      Leo De Hartog   5,083,249 in Books (   

                                           also_view main_cat  Price  \
0                                     ['0590457292']    Books   5.48   
1  ['0385316402', '1890517208', '1933486058', '19...    Books  21.40   
2                                                 []    Books  17.00   
3                                     ['0793506603']    Books  10.95   
4                                                 []    Books   3.50   

         asin                                            details  \
0  B001D4OHQA  {'Publisher:': 'Scholastic (1974)', 'Language:...   
1  B0071QRBFS  {'Paperback:': '400 pages', 'Publisher:': 'Cor...   
2  965906523X  {'Hardcover:': '202 pages', 'Publisher:': 'Fel...   
3  B0011ZV86I  {'Publisher:': 'G. Schirmer, Inc. (1957)', 'AS...   
4  B001LIQC7A  {'Hardcover:': '230 pages', 'Publisher:': 'Bar...   

                                   descriptionstring  
0  Staining on cover, minimal wear and creasing. ...  
1  A double volume with two &quot;front covers.&q...  
2  Stranger in the woods is a dramatic tale of co...  
3             Complete vocal score, words and music.  
4                  a great biography of Ghengis Khan

<div style="border: 4px solid coral; text-align: center; margin: auto;">
    <h3><i>It's time to check your knowledge!</i></h3>
    <br>
    <p style=" text-align: center; margin: auto;">To load the question, run the following cell.</p>
    <br>
</div> 

In [5]:
# Run this cell for a knowledge check question
question_1

---
## Model training with AutoGluon

You can use AutoGluon to train a model by using a single line of code. You need to provide the dataset and tell AutoGluon which column from the dataset you are trying to predict.

<div style="border: 4px solid coral; text-align: center; margin: auto;">
    <h3><i>Try it yourself!</i></h3>
    <br>
    <p style="text-align:center;margin:auto;"><img src="images/activity.png" alt="Activity" width="100" /> </p>
    <p style=" text-align: center; margin: auto;">To prepare the datasets, run the following cell.<br/>
        This step is not required for AutoGluon to work, but it will reduce the time to train your first model.<br/>
The code randomly selects 1,000 rows from the dataset and splits them into training and validation datasets.</p>
    <br>
</div>

In [8]:
# Sampling 1,000
# Try setting the subsample_size to a much larger value to see what happens during training
subsample_size = 1000  # Sample a subset of the data for faster demo
df_train_smaller = df_train.sample(n=subsample_size, random_state=0)

# Print the first few rows
df_train_smaller.head()

category  \
398                                                  []   
3833                                                 []   
4836                             ['Books', 'Reference']   
4572                                                 []   
636   ['Books', 'Arts &amp; Photography', 'Decorativ...   

                                                  title  \
398                       Every Last One (Audiobook CD)   
3833                                          Books" />   
4836  Review Notes and Study Guide to  Conrad's Vict...   
4572                      Simon's Cat va al veterinario   
636         Taisho Kimono: Speaking of Past and Present   

                                               also_buy  \
398   ['1491546336', '1600244041', '1524754668', '14...   
3833  ['0441810764', '0312863551', '0441094996', '04...   
4836                                                 []   
4572                                                 []   
636                                      ['4756246354']   

                                    brand                  rank  \
398   Visit Amazon's -Anna Quindlen- Page  6,392,575 in Books (   
3833                    Robert A Heinlein  4,893,400 in Books (   
4836                            Ken Sobol  2,286,014 in Books (   
4572                        Simon Tofield  7,769,270 in Books (   
636          Visit Amazon's Jan Dees Page  2,053,979 in Books (   

                                              also_view main_cat  Price  \
398   ['0812985907', '0525509879', '0812976185', '08...    Books  23.84   
3833  ['0441810764', '0312863551', '0671577808', '04...    Books   6.74   
4836                                                 []    Books   8.07   
4572                                                 []    Books  15.18   
636                                                  []    Books  51.75   

            asin                                            details  \
398   B003SFS8F8  {'Publisher:': 'Unabridged edition; Unabridged...   
3833  B001R2GZA4  {'Publisher:': 'SIGNET BOOKS (1900)', 'ASIN:':...   
4836  B000QCDE5A  {'Paperback:': '142 pages', 'Publisher:': 'Mon...   
4572  8416261865  {'Publisher:': 'Duomo Ediciones (October 1, 20...   
636   8857200116  {'Hardcover:': '292 pages', 'Publisher:': 'Ski...   

                                      descriptionstring  
398   The latest novel from Pulitzer Prize-winner An...  
3833                     Classic science fiction novel.  
4836                 A CRITICAL GUIDE BY MONARCH NOTES.  
4572                          Brand New. Ship worldwide  
636   A unique collection of 130 kimonos for women, ...

<div style="border: 4px solid coral; text-align: center; margin: auto;">
    <h3><i>It's time to check your knowledge!</i></h3>
    <br>
    <p style=" text-align: center; margin: auto;">To load the question, run the following cell.</p>
    <br>
</div> 

In [9]:
# Run this cell for a knowledge check question
question_2

### Training a model with a small sample

AutoGluon uses certain defaults. For example, AutoGluon uses `root_mean_squared_error` as an evaluation metric for regression problems. For more information, see [sklearn.metrics](https://scikit-learn.org/stable/modules/classes.html#sklearn-metrics-metrics) in the sklearn documentation.

__Note:__ Training on this smaller dataset will take approximately 3–4 minutes.

<div style="border: 4px solid coral; text-align: center; margin: auto;">
    <h3><i>Try it yourself!</i></h3>
    <br>
    <p style="text-align:center;margin:auto;"><img src="images/activity.png" alt="Activity" width="100" /> </p>
    <p style=" text-align: center; margin: auto;">Use `TabularPredictor` to train the first version of the model along with the smaller 1000 sample training dataset so the model trains faster.<br>
</p>
    <br>
</div>

In [10]:
# Run this cell

smaller_predictor = TabularPredictor(label="Price", eval_metric = 'mean_squared_error').fit(train_data=df_train_smaller)

No path specified. Models will be saved in: "AutogluonModels/ag-20230928_033751/"
Beginning AutoGluon training ...
AutoGluon will save models to "AutogluonModels/ag-20230928_033751/"
AutoGluon Version:  0.8.0
Python Version:     3.10.10
Operating System:   Linux
Platform Machine:   x86_64
Platform Version:   #1 SMP Mon Apr 24 23:34:06 UTC 2023
Disk Space Avail:   19.81 GB / 20.96 GB (94.5%)
Train Data Rows:    1000
Train Data Columns: 10
Label Column: Price
Preprocessing data ...
AutoGluon infers your prediction problem is: 'regression' (because dtype of label-column == float and many unique label-values observed).
	Label info (max, min, mean, stddev): (2326.87, 0.0, 39.77738, 123.6481)
	If 'regression' is not the correct problem_type, please manually specify the problem_type parameter during predictor init (You may specify problem_type as one of: ['binary', 'multiclass', 'regression'])
Using Feature Generators to preprocess the data ...
Fitting AutoMLPipelineFeatureGenerator...
	Avail

### Interpreting the training output
AutoGluon outputs a lot of information about what happens during model training.

<div style="border: 4px solid coral; text-align: center; margin: auto;"> 
    <h3><i>Try it yourself!</i></h3>
    <p style="text-align:center; margin:auto;"><img src="images/challenge.png" alt="Activity" width="100" /> </p>
    <p style=" text-align: center; margin: auto;">After the training finishes, examine the output and answer the following questions based on the output.</p>
    <br>
</div>


1. What is the shape of the training dataset?
2. What type of ML problem (such as classification or regression) does AutoGluon infer? (**Hint:** Remember, you didn't mention the problem type. You only provided the label column.)
3. What does AutoGluon suggest in case it inferred the wrong problem type?
4. What kind of data preprocessing and feature engineering did AutoGluon perform?
5. What are the basic statistics about the label in the print statements from AutoGluon?
6. How many extra features were generated in addition to the originals in the dataset? What was the runtime for that?
7. Which evaluation metric was used?
8. What does AutoGluon suggest in case it inferred the wrong metric?
9. What is the ratio between the training and validation dataset? (**Hint:** Look for `val` or `validation`.)
10. Where did AutoGluon save the predictor?
11. Which folder were the models saved in?
12. What file format are the models in? (**Note:** Look at the file name suffix. You don't need to open the file.)

Try to answer these questions before you check the solution.

### List your answers here:
1. 1000 Rows and 10 Columns (1000, 10)

2. AutoGluon infers your prediction problem is: 'regression'

3. It suggests to manually specify the problem_type parameter during predictor init (['binary', 'multiclass', 'regression'])

4. AutoGluon performed various data preprocessing and feature engineering tasks, including: Data type inference for each feature; Fitting feature generators (AsTypeFeatureGenerator, FillNaFeatureGenerator, CategoryFeatureGenerator, TextSpecialFeatureGenerator, TextNgramFeatureGenerator, etc.); BinnedFeatureGenerator, CountVectorizer, and removed unused features.

5. Label info (max, min, mean, stddev): (2326.87, 0.0, 39.77738, 123.6481)
Maximum value: 2326.87
Minimum value: 0.0
Mean: 39.77738
Standard deviation: 123.6481

6. 441 features in addition to the original 9 features in the dataset. The runtime for this feature generation was 10.3 seconds.

7. AutoGluon will gauge predictive performance using evaluation metric: 'mean_squared_error'

8. It mentions that one can change the evaluation metric by specifying the 'eval_metric' parameter of Predictor()

9. Train Rows: 800, Val Rows: 200

10. AutoGluon saved the predictor in the following directory: "AutogluonModels/ag-20230928_033751/"

11. Models were saved in the "AutogluonModels/ag-20230928_033751/" folder

12. The models are saved in .pkl format


----
## Conclusion

The purpose of this notebook was to explore a dataset of information about books and to use AutoGluon to build a basic model to predict book prices based on book features.

## Next lab
In the next lab, you will learn how to use AutoGluon features to refine your model.